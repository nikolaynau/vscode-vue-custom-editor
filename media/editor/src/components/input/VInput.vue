<script lang="ts">
import { defineComponent } from 'vue';

export default defineComponent({
  inheritAttrs: false
});
</script>

<script setup lang="ts">
import { ref, computed, toRefs } from 'vue';

export interface Props {
  modelValue?: string | number;
  type?: string;
  tabIndex?: number;
  className?: string;
  iconClickable?: boolean;
  disabled?: boolean;
  readonly?: boolean;
}

const props = withDefaults(defineProps<Props>(), {
  modelValue: undefined,
  type: 'text',
  tabIndex: 0,
  className: undefined,
  iconClickable: false,
  disabled: false,
  readonly: false
});

const emit = defineEmits<{
  (e: 'update:modelValue', value: string): void;
  (e: 'icon-click', ev: MouseEvent): void;
}>();

const { modelValue, iconClickable, className, disabled, readonly } =
  toRefs(props);
const inputEl = ref(null);

const value = computed<string>({
  get: () => modelValue.value as string,
  set: val => emit('update:modelValue', val)
});

const layoutClasses = computed(() => ({
  [className.value as string]: !!className.value,
  'v-input-layout--disabled': disabled.value,
  'v-input-layout--readonly': readonly.value
}));

const iconClickDisabled = computed(
  () => !iconClickable.value || disabled.value || readonly.value
);

const iconClasses = computed(() => ({
  'v-input__icon--clickable': !iconClickDisabled.value
}));

function onIconClick(e: MouseEvent) {
  emit('icon-click', e);
}

defineExpose({
  inputEl
});
</script>

<template>
  <div class="v-input-layout" :class="layoutClasses">
    <input
      ref="inputEl"
      class="v-input"
      :class="{ 'v-input--icon': $slots.icon }"
      v-model="value"
      v-bind="$attrs"
      :type="type"
      :disabled="disabled"
      :readonly="readonly"
      :tabindex="tabIndex"
    />
    <span
      v-if="$slots.icon"
      class="v-input__icon"
      :class="iconClasses"
      @click="iconClickDisabled ? null : onIconClick($event)"
    >
      <slot name="icon"></slot>
    </span>
  </div>
</template>

<style>
@import 'input';
</style>
