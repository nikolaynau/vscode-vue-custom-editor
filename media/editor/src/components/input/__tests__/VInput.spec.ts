import { describe, it, expect } from 'vitest';
import { mount } from '@vue/test-utils';
import VInput from '../VInput.vue';

describe('VInput', () => {
  it('default render props', () => {
    const wrapper = mount(VInput);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('update model value event', async () => {
    const inputText = 'input text';

    const wrapper = mount(VInput);
    await wrapper.find('input').setValue(inputText);

    const updateModelValueEvent = wrapper.emitted(
      'update:modelValue'
    ) as string[][];

    expect(updateModelValueEvent).toHaveLength(1);
    expect(updateModelValueEvent[0][0] as string).toBe(inputText);
  });

  it('disabled input', () => {
    const wrapper = mount(VInput, { props: { disabled: true } });
    expect(wrapper.classes('v-input-layout--disabled')).toBe(true);
    expect(wrapper.find('input').element.disabled).toBe(true);
  });

  it('readonly input', () => {
    const wrapper = mount(VInput, { props: { readonly: true } });
    expect(wrapper.classes('v-input-layout--readonly')).toBe(true);
    expect(wrapper.find('input').element.readOnly).toBe(true);
  });

  it('type prop', () => {
    const wrapper = mount(VInput, { props: { type: 'checkbox' } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('tabIndex prop', () => {
    const wrapper = mount(VInput, { props: { tabIndex: 1 } });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('custom layout class name prop', () => {
    const wrapper = mount(VInput, {
      props: { className: 'custom-layout-input' }
    });
    expect(wrapper.classes('custom-layout-input')).toBe(true);
  });

  it('input with icon', () => {
    const wrapper = mount(VInput, {
      slots: { icon: 'edit icon' }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('no clickable icon', async () => {
    const wrapper = mount(VInput, {
      slots: { icon: 'edit icon' }
    });

    await wrapper.find('.v-input__icon').trigger('click');
    const iconClickEvent = wrapper.emitted('icon-click');

    expect(iconClickEvent).toBeUndefined();
  });

  it('clickable icon', async () => {
    const wrapper = mount(VInput, {
      props: {
        iconClickable: true
      },
      slots: { icon: 'edit icon' }
    });

    const iconWrapper = wrapper.find('.v-input__icon');
    await iconWrapper.trigger('click');
    const iconClickEvent = wrapper.emitted('icon-click');

    expect(iconClickEvent).toHaveLength(1);
    expect(iconWrapper.classes('v-input__icon--clickable')).toBe(true);
  });

  it('no clickable icon when input is disabled', async () => {
    const wrapper = mount(VInput, {
      props: { disabled: true },
      slots: { icon: 'edit icon' }
    });

    await wrapper.find('.v-input__icon').trigger('click');
    const iconClickEvent = wrapper.emitted('icon-click');

    expect(iconClickEvent).toBeUndefined();
  });

  it('no clickable icon when input is readonly', async () => {
    const wrapper = mount(VInput, {
      props: { readonly: true },
      slots: { icon: 'edit icon' }
    });

    await wrapper.find('.v-input__icon').trigger('click');
    const iconClickEvent = wrapper.emitted('icon-click');

    expect(iconClickEvent).toBeUndefined();
  });
});
