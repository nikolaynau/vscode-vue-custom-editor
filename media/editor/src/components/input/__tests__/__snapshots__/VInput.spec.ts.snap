// Vitest Snapshot v1

exports[`VInput > default render props 1`] = `
"<div class=\\"v-input-layout\\"><input class=\\"v-input\\" type=\\"text\\" tabindex=\\"0\\">
  <!--v-if-->
</div>"
`;

exports[`VInput > input with icon 1`] = `"<div class=\\"v-input-layout\\"><input class=\\"v-input v-input--icon\\" type=\\"text\\" tabindex=\\"0\\"><span class=\\"v-input__icon\\">edit icon</span></div>"`;

exports[`VInput > tabIndex prop 1`] = `
"<div class=\\"v-input-layout\\"><input class=\\"v-input\\" type=\\"text\\" tabindex=\\"1\\">
  <!--v-if-->
</div>"
`;

exports[`VInput > type prop 1`] = `
"<div class=\\"v-input-layout\\"><input class=\\"v-input\\" type=\\"checkbox\\" tabindex=\\"0\\">
  <!--v-if-->
</div>"
`;
